src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceTest.java

```diff
@@ -252,32 +248,138 @@ public void createAiNotebookInstanceUndo() throws Exception {
+    assertNotFound(instanceName, notebooks);
+    assertNotFound(serviceAccountName, crlService.getIamCow());
+    assertThrows(
+        ResourceNotFoundException.class,
+        () ->
+            controlledResourceService.getControlledResource(
+                resource.getWorkspaceId(),
+                resource.getResourceId(),
+                userAccessUtils.defaultUserAuthRequest()));
+  }
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = bufferServiceDisabledEnvsRegEx)
+  public void deleteAiNotebookInstanceUndoIsDismalFailure() throws Exception {
+    ControlledAiNotebookInstanceResource resource =
+        createDefaultPrivateAiNotebookInstance("delete-ai-notebook-instance-undo");
+
+    // Test that trying to undo a notebook deletion is a dismal failure. We cannot undo deletion.
+    jobService.setFlightDebugInfoForTest(
+        FlightDebugInfo.newBuilder().lastStepFailure(true).build());
+    assertThrows(
+        InvalidResultStateException.class,
+        () ->
+            controlledResourceService.deleteControlledResourceSync(
+                resource.getWorkspaceId(),
+                resource.getResourceId(),
+                userAccessUtils.defaultUserAuthRequest()));
+  }
+
+  /** Create a controlled AI Notebook instance with default private settings. */
+  private ControlledAiNotebookInstanceResource createDefaultPrivateAiNotebookInstance(
+      String instanceId) {
+    Workspace workspace = reusableWorkspace();
+    ApiGcpAiNotebookInstanceCreationParameters creationParameters =
+        ControlledResourceFixtures.defaultNotebookCreationParameters()
+            .instanceId(instanceId)
+            .location(DEFAULT_NOTEBOOK_LOCATION);
+    ControlledAiNotebookInstanceResource.Builder resourceBuilder =
+        ControlledResourceFixtures.makeDefaultAiNotebookInstance()
+            .name(instanceId)
+            .workspaceId(workspace.getWorkspaceId())
+            .assignedUser(userAccessUtils.getDefaultUserEmail())
+            .accessScope(AccessScopeType.ACCESS_SCOPE_PRIVATE)
+            .managedBy(ManagedByType.MANAGED_BY_USER)
+            .instanceId(instanceId)
+            .location(DEFAULT_NOTEBOOK_LOCATION);
+    ControlledAiNotebookInstanceResource resource = resourceBuilder.build();
+
+    String createJobId =
+        controlledResourceService.createAiNotebookInstance(
+            resource,
+            creationParameters,
+            DEFAULT_ROLES,
+            new ApiJobControl().id(UUID.randomUUID().toString()),
+            null,
+            userAccessUtils.defaultUserAuthRequest());
+    jobService.waitForJob(createJobId);
+    return jobService
+        .retrieveJobResult(
+            createJobId,
+            ControlledAiNotebookInstanceResource.class,
+            userAccessUtils.defaultUserAuthRequest())
+        .getResult();
+  }
+
+  private static void assertNotFound(InstanceName instanceName, AIPlatformNotebooksCow notebooks) {
```

Comment:
There might be a utility class you can add these to.

Reply:
I think that's premature until they're used by another test class.